### DISCUSSION TITLE
Time estimates for remaining 2.1 work items

### DISCUSSION THREAD

**Discussion by jcnelson** [stacks-network/stacks-core#3096] [General]
## How to Read this Document

All items marked with `Pain level: very high` or `Pain level: high` are _required_ for 2.1.  Everything else is informational.

The time estimations here are in person-hours spent working on the task, where 1 day is 8 person-hours.  These person-hours may be spread across many wall-clock days, depending on whether or not there are other non-2.1 tasks that take precedence.  These estimates do _not_ include time taken to review the code.

The associated Github issues' progress can be tracked on the [Stacks 2.1 project board](https://github.com/stacks-network/stacks-blockchain/projects/11).

## 1-2 days

* Consensus bug: wrong microblock fee calculation [#3140]

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3144
   
* Fix UTXO chaining [#2358]:  This seems to be a 10-line change -- we just need to fix and gate the block height offset calculation.  Most of the work is going to be in testing it.

   * Pain level: high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3110

* Expose chain ID [#2982].  This is the act of adding a new global variable, and gating its usage (including whether or not it's mainnet or testnet)

   * Pain level: high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3111
   
* Expose burnchain tokens spent for a Stacks block via `get-block-info?` or similar, as well as exposing the STX accumulated [#3043] 

   * Pain level: high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3142

* No sortition is possible in the last reward cycle block if the number of reward slots is odd [#2507]:  This is a 10-line change -- we just need to use the correct burn address for mainnet.  Most of the work is going to be in testing it.

   * Pain level: low

* Tolerate flash blocks in PoX prepare phase [#2535]:  This is a matter of changing the burn window size from _w_ to _w - min(M, m)_, where _w_ = 100 per SIP 007, _M_ is the maximum number of missed sortitions we tolerate (e.g. 10), and _m_ is the actual number of missed sortitions.  Most of the work is going to be in testing it.

   * Pain level: low

* Check a contract-call's argument count on analysis [#2683].  This is a 10-line bugfix that must be gated.  It will take more work to test than to implement.

   * Pain level: low
   
* Enforce the right number of arguments for all native types [#3117]

   * Pain level: low

## 2-3 days

* Codec for Clarity values to/from buffers [#2693].  This is part of SIP-018, which will likely be folded into the Stacks 2.1 SIP.  This is a new set of Clarity functions, and would require coordinating with the SIP-018 authors on the semantics.

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3132 
   
* Bitcoin wire format rigidity [#2094], if we take the simple approach of just widening the window for which an on-BTC operation will be treated as valid

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3126

* Extend PoX sunset [#2613].  Just remove the sunset; don't touch `(reject-pox)`; just implement the capped upside for discount-mining

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3129
   
* Require PoX anchor block to correspond to the block commit with the most BTC and confirmations [#2625].  Blocked on merging #2707.

   * Pain level: very high 
   
* Finalize `(at-block)` and `contract-call?` semantics across Clarity versions and epoch boundaries, and add adequate test coverage [#3131]

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3192 

* Implement getters for all state in `.pox-2` and test them [#2538]

   * Pain level: very high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3166

* Treat analysis errors found at runtime as runtime errors [#3107]

   * Pain level: very high
   * https://github.com/stacks-network/stacks-blockchain/pull/3234
 
* Always bill the transaction fee when running a transaction [#3213]

   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3233
   
* Implement getters for the reward set [#3155]

   * Pain level: high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3245
   
* Implement `.costs-3` costs contract [#3115]

   * Pain level: very high
   * https://github.com/stacks-network/stacks-blockchain/pull/3267
   
* Get burn header hash by block height 

   * Pain level: high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/2770
   
* Parse and assemble principals from primitive types

   * Pain level: high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/2756

* Order-independent transaction signing [#2622]:  This requires introducing an additional transaction authentication mode for permitting any order of key signatures, and requires gating its usage until after 2.1 in the block-validation logic and miner.  We'd want to rigorously test this, and we'd probably want to introduce a new transaction version value for this.  AFAIK no hardware wallets even support multisig yet, so we have a somewhat green field to work with here.

   * Pain level: very low

## 3-7 days

* Implement `stacks-unlock` [#2534]

   * Pain level: very high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3260

* Implement `stacks-increase` [#2533]

   * Pain level: very high 
   * https://github.com/stacks-network/stacks-blockchain/pull/3282

* Native segwit support for PoX outputs [#2586].  We'll need to change the type signatures for all `.pox-2` methods that take a PoX address, add support for new PoX address versions in `.pox-2`, and add a variant to `StacksAddress` (i.e. make it an `enum`, not a `struct`) to represent segwit addresses (i.e. witness program version, 32 byte hash).  We'll also need to test this rigorously.

   * Pain level: very high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3283
   
* Update coinbase transaction to pay to contract address [#3101].

   * Pain level: very high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3164

* Cap upside of discount mining [#3095].  This is the act of (a) forcing miners to burn BTC if they spend beyond a given threshold on a block-commit, and (b) figuring out the right way to find the threshold -- i.e. from one or more prepare phases.  This will require some exploration of historic chainstate, and will need some rounds of review.

   * Pain level: very high 
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/3135

* Handle the arrival of a hidden anchor block without reorging the whole chain [#1805]
   * Pain level: very high
   * PR: https://github.com/stacks-network/stacks-blockchain/pull/2707
   
* Compress Clarity contracts and only bill for loading/storing their compressed representation [#2926].  This will require choosing and evaluating a consensus-critical compression algorithm, will require gating the feature, and will require re-assessing the runtime costs in Stacks 2.1 for loading/storing compressed contract data.  We'll need to benchmark the new load/store routines end-to-end.

   * Pain level: medium

**Comment by jcnelson** [comment]
We should 4x these estimates if we want to consider the end-to-end time between writing the code and getting the requite reviewers' sign-offs.

**Comment by diwakergupta** [comment]
@jcnelson I recall you mentioned in this Monday's public blockchain call that Stacks 2.1 work items were on track. I might have missed some details though, so I was wondering if you could provide an update here on 1) what's remaining and 2) rough timeline? Is [this board](https://github.com/stacks-network/stacks-blockchain/projects/11) up-to-date? If so, happy to use that as the reference, though it's hard to get a sense of time estimates from the board.

I think a periodic update on this discussion would also benefit those following this issue but not attending the meetings, and also help inform other stakeholders / community members in case there are places where folks can help chip in to move things along.

Thanks!

**Reply by jcnelson** [reply]
This document answers (1) and (2) to the best of my knowledge, so I'm not sure what else you're asking for here?

This document is edited regularly whenever there's a new PR or a new task to be done.

**Reply by diwakergupta** [reply]
Got it, thanks. I was looking for an estimate for the overall timeline for Stacks 2.1 work being done. Sounds like you're suggesting one could just go item by item and compute an estimate based on whether a PR exists or work has not yet begun, which I can do. Appreciate it!

**Comment by saralab** [comment]
Looking at the current 2.1 Items, the following tasks seem to have the highest impact with the help from team at Hiro. I also, added a note on these issues individually. 

- [https://github.com/stacks-network/stacks-blockchain/issues/3115](https://github.com/stacks-network/stacks-blockchain/issues/3115) : @pavitthrap 
- [https://github.com/stacks-network/stacks-blockchain/issues/2534](https://github.com/stacks-network/stacks-blockchain/issues/2534) - @kantai 
- [https://github.com/stacks-network/stacks-blockchain/issues/2533](https://github.com/stacks-network/stacks-blockchain/issues/2533) - @kantai 
- [https://github.com/stacks-network/stacks-blockchain/issues/3107](https://github.com/stacks-network/stacks-blockchain/issues/3107) - @obycode

**Reply by jcnelson** [reply]
Also @obycode has volunteered to take a crack at implementing more robust trait semantics per https://github.com/stacks-network/stacks-blockchain/discussions/3231#discussioncomment-3300350

**Comment by saralab** [comment]
Based on the estimates here and current WIP, here's the code complete and post code complete timeline. 
Probably optimistic, let's discuss the details during the weekly today. 

Projected Code Complete: Se 09 2022 
Projected 2.1 Activation : Nov 29 2022 


<img width="770" alt="Screen Shot 2022-09-06 at 10 43 49 AM" src="https://user-images.githubusercontent.com/264896/188665284-97ce1353-5ec1-4747-9320-8d892da211e4.png">

**Reply by jcnelson** [reply]
Code-completion is still likely at least a week away, when review times are accounted for.  Maybe as many as three.

**Comment by saralab** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/saralab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/saralab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">



  | Start | End
-- | -- | --
Code   Complete[0] |   | 31-Oct
Private   Testnet [0] |   | 31-Oct
SIP 2.1 to   Advance to Recommended Status [+7] |   | 7-Nov
Public Testnet   [+7] |   | 7-Nov
Integration   Partner Window[0 to +67] | 31-Oct | 30-Nov
Voting Period   [+32 to +67] | 31-Oct | 6-Jan
Stacks 2.1   Binaries [+67] | 1-Dec | 6-Jan
Miners Finish   Upgrading [+81] | 15-Dec | 20-Jan
Stacks 2.1   Activation [+81] | 15-Dec | 20-Jan



</body>

</html>

**Reply by diwakergupta** [reply]
@saralab could you update this with the dates discussed in Monday's call (https://docs.google.com/document/d/1SA87UgCz0S0YK9zpUFUdDCZkLw-QiyizkHhpk6o_OLk/edit) ?

**Reply by saralab** [reply]
Updated based on Monday's Blockchain Weekly discussion

**Reply by diwakergupta** [reply]
Are those "End dates" accurate? I think only start dates were discussed on Monday IIRC
